### PR TITLE
Validation for fakerjs in kafka file

### LINKED ISSUE #154: Validation for fakerjs in kafka file

### PR DESCRIPTION
**Pull request by angelozerr**
Validation for fakerjs in kafka file

Fixes #154

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by fbricon** [comment]
It's broken

https://user-images.githubusercontent.com/148698/115008572-8f8e5c80-9eab-11eb-81a1-bfe7a131e813.mp4

File: src/kafka-file/languageservice/services/diagnostics.ts [hunk 1]
```diff
@@ -31,11 +32,20 @@ export class KafkaFileDiagnostics {
...
+        // 1. Check if producer defines a value content
         const errorMessage = ProducerValidator.validateProducerValue(value?.content);
         if (errorMessage) {
             const range = new Range(block.start, new Position(block.start.line, block.start.character + 8));
             diagnostics.push(new Diagnostic(range, errorMessage, DiagnosticSeverity.Error));
         }
+        // 2. Validate the Faker expressions content
+        value?.expressions.forEach(expression => {
+            const content = expression.content;
+            if (!fakerjsAPIModel.hasDefinition(content)) {
```
Comment (fbricon) [inline src/kafka-file/languageservice/services/diagnostics.ts:44]: Valid expressions break:

Screenshot 2021-04-16 at 12 12 23

File: src/kafka-file/languageservice/services/diagnostics.ts [hunk 2]
```diff
@@ -31,11 +32,20 @@ export class KafkaFileDiagnostics {
...
+                const range = expression.expressionRange;
+                diagnostics.push(new Diagnostic(range, `Invalid Faker expression '${content}'`, DiagnosticSeverity.Error));
```
Comment (fbricon) [inline src/kafka-file/languageservice/services/diagnostics.ts:46]: FakerJS

File: src/kafka-file/languageservice/services/diagnostics.ts (hunk 1 above)
Comment (angelozerr) [reply src/kafka-file/languageservice/services/diagnostics.ts]: fixed

File: src/kafka-file/languageservice/services/diagnostics.ts (hunk 2 above)
Comment (angelozerr) [reply src/kafka-file/languageservice/services/diagnostics.ts]: fixed

**Comment by fbricon** [comment]
Ideally we should prevent nested fakerjs expressions, with a proper message.
Screenshot 2021-04-19 at 11 28 44


Also fakerjs completion shouldn't  be triggered when in a nested expression

**Comment by angelozerr** [comment]
Nested fakerjs expression cannot be hard to manage, because you have the use case when expression is not opened or closed correctly.  I will invesitigate more with this usecase.

**Comment by angelozerr** [comment]
I pushed my work that I did quickly last night. It's a draft. I have no time to explain the strategy now (I will do that tomorow).

Please play with this PR and tell me if you like.

**Comment by angelozerr** [comment]
Here some usecases:

```
PRODUCER keyed-message
topic: my-topic
key: mykeyq
record 

{{address.cardinalDirection}}

{{address.cardinalDirection(}})}}

{{address.caradd{{}}tion}}

{{address.caradd}}tion}}

{{address.caradd{{tion}}

{{address.cardinalDirection{{

{{address.cardinalDirec{{tion}}

```

File: docs/Producing.md [hunk 3]
```diff
@@ -71,10 +71,14 @@ Completion is available for
...
 Validation will help you write valid producers in .kafka files.
 
-Here is an example of value validation:
+ * here is an example of value validation:
 
 ![Empty value](assets/kafka-file-producer-empty-value-validation.png)
 
+ * here is an example with FakerJS validation:
+ 
+![FakerJS validation](assets/ kafka-file-producer-fakerjs-validation)
```
Comment (fbricon) [inline docs/Producing.md:80]: wrong path

**Comment by fbricon** [comment]
I think the validation is good enough.

File: docs/Producing.md (hunk 3 above)
Comment (angelozerr) [reply docs/Producing.md]: fixed

**Comment by angelozerr** [comment]
The PR is ready again. I have fixed tests and add another tests with faker validation (and completion). This PR takes care of producerFakerJSEnabled settings for validation and completion too.

**Comment by fbricon** [comment]
Thanks @angelozerr!

**Pull request merged**